File: sample-java/src/test/java/com/ibotta/gradle/aop/java/AopTest.java
package com.ibotta.gradle.aop.java;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;

public class AopTest {
    @Mock private MessageListener mockMessageListener;
    private JavaTargetExample target = new JavaTargetExample();

    @BeforeEach
    public void before() {
        MockitoAnnotations.openMocks(this);
    }

    @Test
    public void givenAOPBeforeHook_whenTargetInvoked_thenAOPMethodRuns() {
        target.demonstrateJavaAOP(mockMessageListener);
        verify(mockMessageListener).onMessage(anyString(), eq(CallerType.BEFORE_HOOK));
    }

    @Test
    public void givenAOPAfterHook_whenTargetInvoked_thenAOPMethodRuns() {
        target.demonstrateJavaAOP(mockMessageListener);
        verify(mockMessageListener).onMessage(anyString(), eq(CallerType.AFTER_HOOK));
    }
}

File: sample-kotlin/src/test/kotlin/com/ibotta/gradle/aop/kotlin/AopTest.kt
package com.ibotta.gradle.aop.kotlin

import CallerType
import io.mockk.MockKAnnotations
import io.mockk.impl.annotations.MockK
import io.mockk.verify
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class AopTest {
    @MockK(relaxed = true) private lateinit var mockMessageListener: com.ibotta.gradle.aop.MessageListener
    private val target = KotlinTargetExample()

    @BeforeEach
    fun before() {
        MockKAnnotations.init(this)
    }

    @Test
    fun givenAOPBeforeHook_whenTargetInvoked_thenAOPMethodRuns() {
        target.demonstrateKotlinAOP(mockMessageListener)
        verify { mockMessageListener.onMessage(any(), CallerType.BEFORE_HOOK) }
    }

    @Test
    fun givenAOPAfterHook_whenTargetInvoked_thenAOPMethodRuns() {
        target.demonstrateKotlinAOP(mockMessageListener)
        verify { mockMessageListener.onMessage(any(), CallerType.AFTER_HOOK) }
    }
}